stage #0/module #4 Algorithms, Data Structures/algorithms_and_data_structures.md

```diff
@@ -0,0 +1,487 @@
+- "O(n^2)", a.k.a "quadratic time": the number of operations it performs scales in proportion to the square of the input;
+- "O(2^n)", a.k.a "exponential time": any time an input unit increases by 1, it causes you to double the number of operations performed;
+- "O(n!)", a.k.a "factorial time": the number of operations it performs scales in proportion to the result of multiplication of all positive integers less than or equal to the input.
+
+"Big O notation" focuses on the "worst-case" scenario. For example, there is a problem, which is to find a contact in your phone book. We choose the following algorithm: look through the phone book from the very beginning to the end. Let's assume that the number of contacts in the phone book is 100 (n=100), and we are looking for a contact named as "Alexandra". It turned out that "Alexandra" is the first entry in the phone book, so in this case the time complexity will be "O(1)" (we performed a constant number of operation: "1", it will be the "best-case" scenario). Now let's find "Yevgeny", who is the last 100th contact (time complexity will be "O(n)"). This case should be considered as the "worst-case" scenario, when we have to look at each contact in the phone book. Thus, the "worst-case" scenario is a concept describing the maximum time complexity for an algorithm.
+
+### Summary
+- algorithm is a method for solving a problem;
+- algorithm's speed is measured in growth of the number of operations, not in seconds;
+- algorithm's execution time is expressed in "Big O notation".
+
+## Data structures
+### Overview
+Data structures are methods of organizing, processing, retrieving and storing data in a computer system so that operations can be performed upon them more efficiently. More precisely, a data structure is a collection of data values.
+
+There are several basic and advanced types of data structures, all of them are designed to arrange data to match a specific purpose. Data structures allows users to easily access and work with the data they need in appropriate ways. Data structures define the organization of information, in order that both machines and humans can better understand it. 
+
+A data structure may be selected or designed to store data for the purpose of using it with various algorithms. It is not only important to use data structures, but it is also important to choose the proper data structure for each task.
+
+Data structures are divided into two categories:
+- Linear data structure (the elements are arranged in sequence one after the other, for example, arrays, linked lists, etc.);
+
+![Linear Data Structure](img/linear_data_structure.png)
+
+- Non-linear data structure (the elements are arranged in a hierarchical manner where one element will be connected to one or more elements, for example, hash table, tree, etc.).
+
+[comment]: <> (TODO copywriting)
+![Non-linear](img/non_linear_data_structure.png)
+
+### Arrays
+#### Overview 
+An array is a linear data structure:
```

Comment:
maybe adding something like" array is a sequential data structure, which means that this structure consists of data units that follow each other with in some boundaries"